### DISCUSSION TITLE
Enhanced BOM Change Log and Subassembly Change Tracking

### DISCUSSION THREAD

**Discussion by KamilHero** [inventree/InvenTree#7845] [Ideas]
Hi - recently when I was working more with this tool, I noticed some missing functionality which may be very useful to track the changes to the BOM.

### Current Behaviour

When adding parts to the BOM, users have the option to validate the BOM, after which a message is displayed:
"This BOM was last checked by User1 on Aug. 9, 2024".

If any component within the BOM is changed, added, or removed, a message is shown:
"The BOM this part has been changed, and must be validated. This BOM was last checked by User2 on Aug. 8, 2024".

### Requested Enhancements

1. Detailed Change Log with Approval History:

- Implement a detailed change log that tracks all modifications to the BOM, including the following:
  - Approval Date: The date when the BOM was last validated after a change.
  - Approval Person (User): The user who validated or approved the BOM.
  - Change Details: A comprehensive log of what was changed in the BOM, specifying:
    - Components that were added (including details such as part number, quantity, etc.).
    - Components that were removed (including details such as part number, quantity, etc.).
    - Components that were modified (with a summary of what was altered, such as part number changes, quantity adjustments, etc.).
- This log should be easily accessible and provide a clear audit trail for all BOM changes. With this data I can see what changed between some specific approved BOM versions.
2. Improved Subassembly Change Tracking:
- Currently, when a subassembly within a parent BOM is changed and approved, it is no longer highlighted as red in the parent BOM, which can be misleading.
- Implement a feature where subassemblies that have been modified, even after approval, are clearly indicated within the parent BOM until the parent BOM itself is revalidated.
- Introduce a visual indicator (such as a different color or icon) that signifies that while the subassembly has been approved, the parent BOM has not yet been revalidated to reflect these changes.
- This will ensure that users can easily identify all changes within a BOM, including those in subassemblies, and take appropriate actions.

Please let me know what you think about it - thanks!

**Comment by matmair** [comment]
This request sounds more or less like detailed revision tracking and change management, which is quite a feat. We added basic revisions in https://github.com/inventree/InvenTree/issues/5028 - that will be in 0.16.0.

**Comment by SchrodingersGat** [comment]
@KamilHero there are some other historical issues where we have discussed this:

- https://github.com/inventree/InvenTree/issues/377
- https://github.com/inventree/InvenTree/issues/4468

As @matmair raises, this would be more of an overall change tracking system, which would require significant investment to implement.

We *currently* track history of stock items with specific operations, which is different to a generic "change history" tracking.

And both of these are separate to what you are asking for here.

A "BOM change tracking" system requires a lot of very specific data to be captured and categorised - it goes well beyond just "this record changed in the database". Further, it is tough to highlight changes which may occur "deep down" in a multi-level bill of materials.

One consideration - we now have the ability to "lock" a BOM, which nominally prevents any further changes. It may be simpler to consider locking down a BOM, and then making a new revision of the assembly. Then, you can compare the BOMs between the two revisions as a clearer way to see differences?

I'm not *against* such a change tracking system - but I cannot envision how it could be implemented in a non-trivial way.
